### ISSUE TITLE
.keras model with base model trainable has poor performance in TensorFlow 2.17 (Keras 3)

### ISSUE THREAD

**Issue by nkinnaird** [keras-team/keras#20147]
I originally posted this bug in the [TensorFlow github issues section](https://github.com/tensorflow/tensorflow/issues/74170) since I believed it to be due to a higher TF version, but was asked to post here since it may instead be due to Keras 3. I am copying my post below:

I am training an EfficientNet model with a custom head using TensorFlow and Keras, saving the model to a .keras format. If the base model trainable flag is set to False, such that I only train the head, then when I later load the .keras model and evaluate it on a dataset, I get the expected good performance. When I set the trainable flag to True and train a model (which converges well), then when I later load the model and evaluate it on the same dataset the performance has degraded significantly. (I am evaluating the model on the same dataset using the same code both at the end of training, and later on in a separate notebook. It is in this separate notebook where the performance is bad, where again the same dataset is being used and the same code is being used in both evaluation places.)

Saving to a .h5 model does not have this issue, and the performance of the saved model is good. I have spent the day trying different trainable and training flag values in various places to no improvement, thinking originally that it was something to do with the BatchNorm layers in the model. Recompiling the model has not helped.

When I switch back to an older TensorFlow version (2.15.0.post1) with Keras 2 I do not see this issue. Both the trained .keras and .h5 models perform well when later loaded and evaluated on my dataset of interest.

This seems like a bug to me, though I also acknowledge that perhaps I have missed something in the TF/Keras updates. I have searched the TensorFlow API docs for the various methods to no success. If it is the latter I would be very grateful for any advice, thank you.

**Comment by james77777778** [comment]
It seems there is a bug in the `EfficientNet` family.

Here is the reproducible script:

```python
import numpy as np

import keras

# Build custom model from EfficientNetB0.
img_size = 224
num_classes = 10
inputs = keras.Input((img_size, img_size, 3))
base_model = keras.applications.EfficientNetB0(
    weights="imagenet", include_top=False
)
x = base_model(inputs)
x = keras.layers.Dense(512, activation="relu")(x)
x = keras.layers.BatchNormalization()(x)
x = keras.layers.Dropout(rate=0.5)(x)

# Another fully connected dense layer
x = keras.layers.Dense(256, activation="relu")(x)
x = keras.layers.BatchNormalization()(x)
x = keras.layers.Dropout(rate=0.5)(x)

outputs = keras.layers.Dense(num_classes, activation="softmax")(x)
model = keras.Model(inputs, outputs)
weights = model.get_weights()

# Verify weights.
model.save("model.keras")
revived_model = keras.models.load_model("model.keras")
revived_weights = revived_model.get_weights()

np.testing.assert_equal(len(weights), len(revived_weights))
for w, rw in zip(weights, revived_weights):
    np.testing.assert_allclose(w, rw)
print("Weights are all the same.")

# Predict.
images = np.random.uniform(size=(2, 224, 224, 3)).astype("float32")
outputs = model.predict(images)
revived_outputs = revived_model.predict(images)
np.testing.assert_allclose(outputs, revived_outputs)  # This line failed.
print("Predictions are all the same.")

```

However, the script works fine with:
- `MobileNet*`
- `EfficientNetV2`

The root cause might be in `keras.applications.EfficientNet` since the saving / reloading work fine with other models.

**Comment by fchollet** [comment]
The issue has to do specifically with the rescaling layer that gets added when loading ImageNet weights:


```python
x = layers.Rescaling([1.0 / math.sqrt(stddev) for stddev in IMAGENET_STDDEV_RGB])(x)
```

**Comment by Doch88** [comment]
So, I investigated it a bit and the problem is the keras.layers.Normalization layer, which is not finalized.
This is because keras.layers.Normalization has two weights that are loaded from file adapted_mean and adapted_std, but this are not used directly when making a prediction.
What is used directly are two variables, which are not added as weights, called mean and std, that are updated whenever finalize_state() is called.

And this is the problem. Theoretically, whenever loading weights from file, finalize_state() should be called by load_own_variables().
But load_own_variables() is called only with the new saving API, when the weights are saved in the format ".weights.h5", while the efficientnet weights have the legacy extension ".h5".

Therefore, weights are loading, we have adapted_mean and adapted_std, but the actual mean and std are the one initialized with build(), i.e. a vector of zeros and a vector of ones.

As a workaround, simply calling `base_model.layers[2].finalize_state()` will fix the issue.

**Comment by Doch88** [comment]
I proposed my fix here: https://github.com/keras-team/keras/pull/20163
